### DISCUSSION TITLE
Reticulum over Avian Carriers

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#617] [General]
I saw this funny IP over Avian Carriers page and it reminded me of reticulum. https://en.m.wikipedia.org/wiki/IP_over_Avian_Carriers

Any ideas for how to leverage our bird friends? Attach nodes to them? Send QR code encrypted messages attached to their legs?

**Comment by markqvist** [comment]
Now that [interface modularity](https://markqvist.github.io/Reticulum/manual/interfaces.html#custom-interfaces) has been implemented, I highly encourage you to write an `AvianCarrierInterface` module!

A semi-automated approach is probably the most viable option here, and one could imagine something like a receipt printer spitting out packets encoded as QR-codes on the transmit side, with something simple robotics rolling up paper strips into small container, that can be magnetically attached to the legs of our pigeon friends, and then unlatching the cage of the bird in question.

On the receiving end, a similar approach could be taken, with some sort of semi-automated feeding mechanism for received paper strips, attached to a QR-code scanner.

In most scenarios though, it's probably worthwhile to consider that rather substantial battery of birds is required. The amount of bird-feed needed, personnel for voliere cleaning, and other physical upkeep requirements are probably also things that are essential to factor into the overall system design.

**Reply by trosel** [reply]
> a receipt printer spitting out packets encoded as QR-codes on the transmit side

unrelated to birds, but I thought long and hard about a physical printer messenger product a couple of weeks ago. Something simplistic like a shared telephone. One printer per household, etc.

**Reply by joakim** [reply]
I imagine an old robed maester handing thermal printed rolls of binary data to an apprentice, saying "release the ravens" :)

I have been wondering if there's an alternative to QR codes for binary data with _arbitrary_ resolution/length, not just a square. Something like [Data Matrix](https://en.wikipedia.org/wiki/Data_Matrix) with arbitrary length that's optimized for encoding an 8-bit bytestream. Essentially like punched tape only as a black/white matrix. It should only encode octets – self-describing CBOR/MessagePack data, UTF-8 text, ciphertext etc. Perfect for thermal printers.

Although not square, image processing should be capable of reading a "strip" by moving the camera along it. Or the other way around, a display could scroll the strip for a camera to scan it. Not that the strip would be very long in most cases. Linear scanners could also work. Does something like this exist?

**Reply by faragher** [reply]
PDF 417 is twice as bulky as QR, but much more arbitrary in size: https://en.wikipedia.org/wiki/PDF417

I've never actually looked into the capacity, but the MDU should be 750, if I'm reading Wikipedia correctly.

I use it for labeling storage boxes and both the USPS and DMV use it quire regularly. It's not *quite* as widely supported on consumer devices, but anything commercial-grade will read them with ease.

There may be better options, but this is a very solid one, even if QR can be more physically compact. I like to think of it as a supported alternative rather than the one and only transfer medium.

**Reply by faragher** [reply]
> Although not square, image processing should be capable of reading a "strip" by moving the camera along it. Or the other way around, a display could scroll the strip for a camera to scan it. Not that the strip would be very long in most cases. Linear scanners should also work. Does anything like this exist?

I mean, you're basically describing a paper tape. Doing so with lines instead of holes is likely easier these days, but I'd start my investigations there.

**Reply by joakim** [reply]
PDF 417 would probably work, but it's very inefficient compared to QR and Data Matrix, as far as I can tell because it has to support handheld linear scanners. I suppose it's rock solid though.

But yes, essentially a paper tape with something like Data Matrix or PDF 417 instead of punched holes. I'm thinking with the compactness of Data Matrix and the linearity of PDF 417. It seems like an interesting challenge. Unfortunately, I don't have the time to tackle it.

I had the idea when staring at my mini thermal printer, wondering how I could use it for Reticulum. It used to be an SMS printer, so I'll probably have it print LXMF messages. I just love receiving small paper telegrams.

**Reply by joakim** [reply]
Speaking of paper tape and encryption 😳
 
> During the last third of the 20th century, the National Security Agency (NSA) used punched paper tape to distribute cryptographic keys. … NSA has been trying to replace this method with a more secure electronic key management system, but as of 2016, paper tape was apparently still being employed.

https://en.wikipedia.org/wiki/Punched_tape#Cryptography

I wonder if they used Avian Carriers.

**Reply by faragher** [reply]
Well, as an exercise, let's try to make a standard.

On the physical level, let's assume you want an eight bit width (16, 32, or even 64 may make more sense depending on block size). Relying on constant speed is problematic, so we'll put a clock track in as well. Since we don't need it for clock, just timing, we'll use off/on instead of individual holes. We can't assume the traction is perfect, so we'll use a second clock track with two on, two off, and use the two track clocks as a linear encoder. This way we can detect missed steps and reversed motion. The use of even numbers of bytes only, or unique stop/start data, can support reading a tape backwards and still get the proper message out.

This leads to a 10 bit width track for 8 data bits. That's probably all we need for a physical format.

From a data structure perspective, you'd want error correction. Hamming is the simplest solution, and using any of their systems would function well, with around 25% of the message being taken up by parity bits. By interleaving the blocks, you can avoid issues with multiple damages bits per block, which is quite likely on a physical tape. Things get quite out of sync rather quickly on the tape, but this may be a feature, not a bug.

OTOH, you can use Hamming(15,11) to align all the bits. With a 15 track tape, you can have 8 data bits, 2 clock bits, and one extra (I can't think of a good use at this time. End marker? Contextual? Third clock track?) and have *everything* have a checsum so you'd know if data or clock tracks were damaged. Increased block size doesn't produce nearly as nice of a size, so mayve we go back and just declare our tape's physical medium 15 tracks and be done with it.

I think it's legitimately clever, but my lunch is done, so my random think time is over. :D

**Reply by joakim** [reply]
> Well, as an exercise, let's try to make a standard.

Haha, OK!

> We can't assume the traction is perfect, so we'll use a second clock track with two on, two off, and use the two track clocks as a linear encoder. This way we can detect missed steps and reversed motion.

I thought one clock/timing track would do, but it's well worth having a second track for those reasons. Nice! I'm thinking they should be the two outer tracks, as delimiters and to indicate up/down.

> From a data structure perspective, you'd want error correction. Hamming is the simplest solution … OTOH, you can use Hamming(15,11) to align all the bits.

I was initially thinking Reed–Solomon which QR, Data Matrix and PDF 417 uses. But your suggestion is much simpler, very elegant and clever! Good luck flipping bits with a marker :)

I like this standard, nicely done! Thanks for running with it :)

**Reply by faragher** [reply]
> I thought one clock/timing track would do, but it's well worth having a second track for those reasons. Nice! I'm thinking they should be the two outer tracks, as delimiters and to indicate up/down.

Oh! That's clever. I hadn't considered that the timing track wasn't symmetrical and would tell you which end was up. It could be anywhere, but at the top and bottom would make it much more human readable. 

I just realized we can't actually interleave the hamming in the system I described, but that's something for a really well-considered system, plus you'd probably want more robust error correction, anyway. It's still far more robust than original paper tapes, though. That was fun.

**Reply by joakim** [reply]
It was! True, the timing tracks would indicate up/down regardless of position. Having them on either side doubles as delimiters and would also add some degree of redundancy.

Since we're not writing an RFC here, I don't think I'll go farther down this rabbit/black hole :)

But if I were to 😁, I'd try to make the bandwidth – literally the width of the band on paper – adjustable. With interleaving error-corection, the Hamming code block size could also be adjustable. One should be able  to pick `(22, 16)` (27% overhead), `(39, 32)` (18%), `(72, 64)` (11%) or even `(137, 128)` (only 7%). The timing tracks would be left out of the Hamming code in this system. If only I had more time.. (Edit: Oh, that's basically what your first suggestion did.)

**Comment by faragher** [comment]
Might I propose rewritable RFID tags, attached or implanted? It doesn't
help the battery of birds problem, but it does automate the I/O process.

**Reply by markqvist** [reply]
Now we're talking - you're bringing the big brain to the discussion here. Rewritable RFID tags is of course the way to go.

**Reply by joakim** [reply]
Yes, now we're talking!

https://www.asiarfid.com/rfid-pigeon-rings.html

Not spam, I promise :)

**Comment by KenSamson** [comment]
![Screenshot_20250205_154239_X](https://github.com/user-attachments/assets/6e4fce8b-6ddc-4665-9a80-1f8a66b0186c)

**Reply by LinuxinaBit** [reply]
I don’t know why, but seeing that somebody took the time to generate and post a needless AI generated image here has made my day considerably worse.

**Reply by virtadpt** [reply]
It doesn't even scan.
